### PR TITLE
Add new JetBrains checker

### PR DESCRIPTION
**Pull request by Lctrs**
Missing tests

### CODE REVIEW

**Comment by Lctrs** [comment]
Kudos to @x80486 for the original code :)

**Comment by Lctrs** [comment]
Ping @FakeShemp and @johanfleury. As maintainers of flatpak wrappers for some Jetbrains products, you might find some interest in this.

**Comment by Lctrs** [comment]
@wjt ready for review :pray:

**Comment by Lctrs** [comment]
Test result is here : https://github.com/Lctrs/flatpak-external-data-checker/commit/68ec6a40b6306b4e89eb85c1209f19e46103d963/checks?check_suite_id=343482923

File: src/checkers/jetbrainschecker.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class JetBrainsChecker(Checker):
+    def check(self, external_data):
+        if not _should_check(external_data):
+            log.debug("%s is not a jetbrains type ext data", external_data.filename)
+            return
+
+        code = external_data.checker_data.get("code")
+
+        assert code is not None
```
Comment (wjt) [inline src/checkers/jetbrainschecker.py]: You can just write

```python3
code = external_data.checker_data["code"]
```

File: src/checkers/jetbrainschecker.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        url = Template("https://data.services.jetbrains.com/products/releases?code=$code&latest=true&type=$type")\
+            .substitute(
+                code=code, type=external_data.checker_data.get("release-type", "release")
+            )
```
Comment (wjt) [inline src/checkers/jetbrainschecker.py]:
```python3
release_type = external_data.checker_data.get("release-type", "release")
url = f"https://data.services.jetbrains.com/products/releases?code={code}&latest=true&type={type}"
```

File: src/checkers/jetbrainschecker.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        data = json.load(resp)[code][0]
+        release = data["downloads"]["linux"]
+
+        checksum = urllib.request.urlopen(release["checksumLink"]).read().decode("utf-8").split(" ")[0]
+
+        new_version = ExternalFile(
+            release["link"], checksum, release["size"], data["version"], data["date"]
+        )
+
+        external_data.state = ExternalData.State.VALID
```
Comment (wjt) [inline src/checkers/jetbrainschecker.py]: This is meant to refer to the state of the _current_ version, not of the new version. I think you can just drop this line.

File: tests/test_jetbrainschecker.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            b"dbb7bf79750b476d2c96e76be6c6082d63c2addffd2694d4d9fa9d6db01162a2 *PhpStorm-2019.3.tar.gz",
+            200
+        )
+
+    return MockResponse(None, 404)
+
+
+class TestJetBrainsChecker(unittest.TestCase):
+    def setUp(self):
+        logging.basicConfig(level=logging.DEBUG)
```
Comment (wjt) [inline tests/test_jetbrainschecker.py]: use init_logging() from utils (see c7f0a7cd1591a4d9b03d78f751566a7942f1935f)

**Review by wjt — CHANGES_REQUESTED** [review]
This is a great idea, thanks for the contribution! Just a few bits and bobs.

File: src/checkers/jetbrainschecker.py (hunk 3 above)
Comment (Lctrs) [reply src/checkers/jetbrainschecker.py]: ~~Looks like this line is needed ? :thinking: The test is failing now.~~

File: src/checkers/jetbrainschecker.py (hunk 2 above)
Comment (Lctrs) [reply src/checkers/jetbrainschecker.py]: I'm not a Python developer (in fact, those are my first Python lines :tada:), but I read on this [blog post](https://realpython.com/python-string-formatting/) that String interpolation in Python are not safe from user input, while Template Strings are. I'm concerned that a malicious user can just get @flathubbot's token access in the log of the action for example.

Comment (barthalion) [reply src/checkers/jetbrainschecker.py]: Technically speaking, I can think of multiple ways a malicious user could already take and none includes string interpolation… So I wouldn't worry too much.

Comment (Lctrs) [reply src/checkers/jetbrainschecker.py]: Okay done.

**Review by wjt — APPROVED** [review]
LGTM, thanks!

**Pull request merged**